### ISSUE TITLE
Avoid storing both coffee and js files in repository

### ISSUE THREAD

**Issue by simi** [teampoltergeist/poltergeist#745]
Problems:
1. mess in code and PR
2. different output on every platform / system / coffee version

To avoid this we can:
1. move to plain JS (whatever version we will decide)
2. compile coffee to JS locally, but don't version JS files (and just add them to released gems) _this probably will block using repository as bundler git dependency_
3. ?

I think move to plain JS make sense here, but it will need a little refactoring since complex constructs like `extends` are used in coffee files. My initial idea was to use ES6, but it is not supported natively by PhantomJS.

/cc @teampoltergeist/committers

**Comment by route** [comment]
For 2) you mean to compile coffee to js?

**Comment by simi** [comment]
Yup. Fixed.

**Comment by twalpole** [comment]
I don't think moving to plain js makes any sense here -- if we can't even have the features of ES6 then it really would be a move backwards and there is nothing to justify it. 

The issue of different output from different versions of coffeescript *should* be fixable by locking the version of coffeescript required, although even then what is the real issue with different JS output as long as it passes the test suite? 

Having both the JS and coffee in the repo - whats the negative other than a little bit of extra storage.  It's not like people are offering up PRs with the JS edited instead of the Coffeescript.

**Comment by route** [comment]
Moving to ES6 was reasonable if PhantomJS would have support it, otherwise we are in the same situation.

**Comment by route** [comment]
I still remember we got commits in master that were fixing coffee, but js files were untouched ;)

**Comment by twalpole** [comment]
@route - yeah -- but the JS gets compiled for the test runs - so any coffee changes are being tested - and master would be no worse than it was before the PR - as long as the files get compiled before any gem is released there is no harm done.

**Comment by simi** [comment]
Yes, at first I thought it is supported. Actually it is not a big deal to transform current coffee syntax into JS also with "advanced" constructs.

My rationale behind this idea is to keep it clean and straightforward. There's no reason to keep `source` and `compiled` files as part of the source code.

For example this example [commit](https://github.com/teampoltergeist/poltergeist/commit/f9ae30b872eb98b99b5e7caa77cc789f663dd4f3). You need to check coffee changes, also compile them in head into JS and check those. Or alternatively you can just try to compile those files locally and compare to those commited one. Also you can force CI to check for changes. But as was written before - for different systems you can get different results.

**tl;dr** - bigger change == bigger headache

**Comment by simi** [comment]
Also I forgot to mention the opposite way pains. When you're debugging some issue, you're getting stacktrace for JS files, not for coffee files.

**Comment by simi** [comment]
Also I can share a short story. Yesterday I was debugging some problem in JS part and I had to do some tracking in compiled js files. Since this whole process is anoying:

1. find JS line in stacktrace
2. find it *manually* in source coffee file
3. change it
4. compile
5. test
6. back to 1 if not solved

Everytime I end up with this:

1. find JS line in stacktrace
2. change line in compiled JS line
3. test
4. back to 1 if not solved
5. move that change to source coffee file
  1. compile
  2. test again

Ideal process should be IMHO:

1. find JS line in stacktrace
2. change line in compiled JS line
3. test
4. back to 1 if not solved

**Comment by route** [comment]
@teampoltergeist/committers I heard PhantomJS will support es6 soon. Thoughts on this?

**Comment by twalpole** [comment]
Yes - PhantomJS 2.5 supports ES6/2015 and Coffeescript 2 just released it's first alpha which generates ES2015 output.  If we're willing to go PhantomJS 2.5 only (once it stabilizes) then we could support those, but if we still want to support PhantomJS 2.1 and 2.5 then we'd need to run through Babel or equivalent to output ES3 for PhantomJS 2.1 - or (the easier option) just stay with the coffeescript we have now which will work on both.

**Comment by johnlpollard** [comment]
Is there a way to use PhantomJS 2.5 or Poltergeist with ES6 now? If so, how?

**Comment by twalpole** [comment]
@johnlpollard PhantomJS 2.5 is still very much an alpha and has been abandoned, you can use it with Poltergeist but there are some broken things. There is work ongoing to produce a Poltergeist compatible driver that uses Puppeteer rather than PhantomJS but it is not in a releasable state yet.
